edge-modules/TwinTester/Program.cs

```
@@ -0,0 +1,344 @@
+            try
+            {
+                await moduleClient.UpdateReportedPropertiesAsync(twin);
+            }
+            catch (Exception e)
+            {
+                string failureStatus = $"{(int)StatusCode.ReportedPropertyUpdateCallFailure}: Failed call to update reported properties";
+                Logger.LogError(failureStatus + $": {e}");
+                await CallAnalyzerToReportStatus(analyzerClient, Settings.Current.ModuleId, failureStatus, string.Empty);
+                return;
+            }
+
+            try
+            {
+                await storage.AddReportedPropertyUpdate(reportedPropertyUpdateCounter.ToString());
+                reportedPropertyUpdateCounter += 1;
+            }
+            catch (Exception e)
+            {
+                Logger.LogError($"Failed adding reported property update to storage: {e}");
+                return;
+            }
+        }
+
+        // TODO: document somewhere that this method could fail due to too short interval of twin updates not giving edgehub enough time
+        static async Task PerformTwinTestsAsync(RegistryManager registryManager, ModuleClient moduleClient, AnalyzerClient analyzerClient, Storage storage)
+        {
+            await ValidateDesiredPropertyUpdates(registryManager, moduleClient, analyzerClient, storage);
+            await ValidateReportedPropertyUpdates(registryManager, moduleClient, analyzerClient, storage);
+            await PerformDesiredPropertyUpdate(registryManager, storage);
+            await PerformReportedPropertyUpdate(moduleClient, analyzerClient, storage);
+        }
+
+        static async Task OnDesiredPropertyUpdateAsync(TwinCollection desiredProperties, object userContext)
+        {
+            // TODO: If expected behavior is calling once per desired property update, then we should not be looping
+            Storage storage = (Storage)userContext;
+            foreach (dynamic twinUpdate in desiredProperties)
```

I tried casting in the loop, but silently continued without iterating.